### PR TITLE
WIP: Fix reduce accumulation for JIT

### PR DESCRIPTION
**Pull request by kvkenyon**
Reduce (non-full) still accumulates into the output buffer. In JIT this causes gradients to be miscomputed as the buffers are not cleared via zero_grad.

I added full-axis with local accumulation. Others might be a bit tricky without adding some weird conditional.

Added a deepwalk and tensor analysis (basic). This let's you see the gradients and their addresses associated to each tensor.

Clang now suports accumulation and non-pointer locals with float or int.

Added a new test case to test_nn that fails due to JIT gradient explosion

### CODE REVIEW

**Comment by kvkenyon** [comment]
The explosion of gradients is fixed by forcing native code to overwrite memory before any stores. Reduce was violating this by not using an accumulator but the value previously stored in the buffer. Updated autograd logic respects the requires_grad flag. This ensures the minimum amount of computation. The final problem is that assigns (due to using Tensor.replace(..)) are not being JIT compiled. Specifically an assign to a realized parameter in optimizer.step() will not mutate unless we still rely on the python execution environment.
